File: ios/background_beacon_sdk/Sources/background_beacon_sdk/BackgroundBeaconSdkPlugin.swift
import Flutter
import UIKit

/**
 * iOS entry point — class name must match `pluginClass` in pubspec.yaml.
 *
 * Job: receive channel calls → convert Map ↔ struct → delegate to
 * BeaconMonitor. No CoreLocation logic here (that lives in BeaconMonitor).
 *
 * Relaunch from a region event (app killed): no launchOptions handling
 * needed — GeneratedPluginRegistrant calls register as the engine comes up,
 * so BeaconMonitor exists with its delegate immediately and CL delivers the
 * pending event on its own. Events arriving before Dart starts listening
 * are buffered and flushed in onListen.
 */
public class BackgroundBeaconSdkPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private let monitor = BeaconMonitor()
    private let permissions = PermissionManager()

    private var eventSink: FlutterEventSink?

    private var pendingEvents: [[String: Any]] = []

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = BackgroundBeaconSdkPlugin()

        let methodChannel = FlutterMethodChannel(
            name: "background_beacon_sdk/methods",
            binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(
            name: "background_beacon_sdk/events",
            binaryMessenger: registrar.messenger())

        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    override init() {
        super.init()
        monitor.onEvent = { [weak self] event in
            self?.send(event)
        }
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "detectMobileServices":
            result("ios")

        case "requestPermissions":
            permissions.request(result: result)

        case "startMonitoring":
            startMonitoring(call, result: result)

        case "stopMonitoring":
            monitor.stopMonitoring()
            result(nil)

        case "detectBeacon":
            guard let args = call.arguments as? [String: Any?],
                  let region = BeaconRegionData(map: args)
            else {
                result(FlutterError(
                    code: "DETECT_FAILED", message: "Invalid region", details: nil))
                return
            }
            monitor.detectBeacon(region: region, timeoutMs: Self.detectTimeoutMs) { found in
                result(found)
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func startMonitoring(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any?],
              let regionMaps = args["regions"] as? [[String: Any?]],
              let settingsMap = args["settings"] as? [String: Any?],
              let settings = ScanSettingsData(map: settingsMap)
        else {
            result(FlutterError(
                code: "START_FAILED", message: "Invalid arguments", details: nil))
            return
        }

        guard !regionMaps.contains(where: { !($0["uuid"] is String) }) else {
            result(FlutterError(
                code: "START_FAILED",
                message: "Wildcard region (uuid == null) is not supported on iOS — "
                    + "CLBeaconRegion requires a known UUID",
                details: nil))
            return
        }

        let regions = regionMaps.compactMap(BeaconRegionData.init(map:))
        guard regions.count == regionMaps.count else {
            result(FlutterError(
                code: "START_FAILED", message: "Invalid region in list", details: nil))
            return
        }

        do {
            try monitor.startMonitoring(regions: regions, settings: settings)
            result(nil)
        } catch {
            result(FlutterError(
                code: "START_FAILED", message: error.localizedDescription, details: nil))
        }
    }

    private func send(_ event: [String: Any]) {
        guard let sink = eventSink else {
            pendingEvents.append(event)
            if pendingEvents.count > Self.pendingEventLimit {
                pendingEvents.removeFirst()
            }
            return
        }
        sink(event)
    }

    // MARK: - FlutterStreamHandler

    public func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        eventSink = events
        pendingEvents.forEach(events)
        pendingEvents.removeAll()
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    private static let detectTimeoutMs = 3000
    private static let pendingEventLimit = 100
}

File: android/src/main/kotlin/com/example/background_beacon_sdk/BackgroundBeaconSdkPlugin.kt
package com.example.background_beacon_sdk

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import com.example.background_beacon_sdk.core.BeaconRegionData
import com.example.background_beacon_sdk.core.BeaconScanner
import com.example.background_beacon_sdk.core.BeaconStore
import com.example.background_beacon_sdk.core.ScanSettingsData
import com.example.background_beacon_sdk.gms.GmsBeaconScanner
import com.example.background_beacon_sdk.hms.HmsBeaconScanner
import com.example.background_beacon_sdk.util.MobileServicesDetector
import com.example.background_beacon_sdk.util.PermissionManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

/**
 * Android entry point — class name must match `pluginClass` in pubspec.yaml.
 *
 * Job: receive channel calls → convert Map ↔ data class → delegate to the
 * scanner. No scan logic here (that lives in core/).
 */
class BackgroundBeaconSdkPlugin :
    FlutterPlugin,
    MethodCallHandler,
    EventChannel.StreamHandler,
    ActivityAware {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context

    private val mainHandler = Handler(Looper.getMainLooper())
    private val permissionManager = PermissionManager()

    private var activity: Activity? = null
    private var activityBinding: ActivityPluginBinding? = null
    private var eventSink: EventChannel.EventSink? = null
    private var scanner: BeaconScanner? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        // detect/scan use applicationContext — no activity needed
        // (an activity is required only for requestPermissions)
        context = binding.applicationContext

        methodChannel = MethodChannel(
            binding.binaryMessenger,
            "background_beacon_sdk/methods",
        )
        eventChannel = EventChannel(
            binding.binaryMessenger,
            "background_beacon_sdk/events",
        )

        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
    }

    /** Create the scanner from detect on first use, then reuse for the engine's lifetime */
    private fun ensureScanner(): BeaconScanner =
        scanner ?: when (MobileServicesDetector.detect(context)) {
            "hms" -> HmsBeaconScanner(context)
            else -> GmsBeaconScanner(context)
        }.also { created ->
            // Scan callbacks arrive on a binder thread — eventSink must be called on main
            created.setListener { event ->
                mainHandler.post { eventSink?.success(event.toMap()) }
            }
            scanner = created
        }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "detectMobileServices" ->
                result.success(MobileServicesDetector.detect(context))

            "requestPermissions" -> {
                val currentActivity = activity
                    ?: return result.error(
                        "NO_ACTIVITY",
                        "requestPermissions requires a foreground activity",
                        null,
                    )
                permissionManager.request(currentActivity, result)
            }

            "startMonitoring" -> {
                try {
                    val regions = call.argument<List<Map<String, Any?>>>("regions")!!
                        .map { BeaconRegionData.fromMap(it) }
                    val settings =
                        ScanSettingsData.fromMap(call.argument<Map<String, Any?>>("settings")!!)
                    ensureScanner().startMonitoring(regions, settings)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("START_FAILED", e.message, null)
                }
            }

            "stopMonitoring" -> {
                scanner?.stopMonitoring()
                result.success(null)
            }

            "detectBeacon" -> {
                try {
                    @Suppress("UNCHECKED_CAST")
                    val region = BeaconRegionData.fromMap(call.arguments as Map<String, Any?>)
                    // Callback already arrives on main (BeaconScanner.detectBeacon's contract)
                    ensureScanner().detectBeacon(region, DETECT_TIMEOUT_MS) { found ->
                        result.success(found)
                    }
                } catch (e: Exception) {
                    result.error("DETECT_FAILED", e.message, null)
                }
            }

            "registerBackgroundCallback" -> {
                // The channel delivers small ints as Int, large as Long — accept via Number
                val dispatcherHandle =
                    (call.argument<Number>("dispatcherHandle"))?.toLong()
                val callbackHandle = (call.argument<Number>("callbackHandle"))?.toLong()
                if (dispatcherHandle == null || callbackHandle == null) {
                    result.error("REGISTER_FAILED", "Missing callback handles", null)
                } else {
                    BeaconStore.saveCallbackHandles(context, dispatcherHandle, callbackHandle)
                    result.success(null)
                }
            }

            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        activityBinding = binding
        binding.addRequestPermissionsResultListener(permissionManager)
    }

    override fun onDetachedFromActivity() {
        activityBinding?.removeRequestPermissionsResultListener(permissionManager)
        activityBinding = null
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        scanner?.stopMonitoring() // engine gone — the scan must not leak and keep draining battery
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    private companion object {
        const val DETECT_TIMEOUT_MS = 3000L
    }
}
